aten/src/ATen/core/LazyTensor.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include <c10/core/TensorImpl.h>
+#include <ATen/core/ivalue.h>
+
+namespace at {
+
+struct CAFFE2_API LazyTensorImpl : public TensorImpl {
+  explicit LazyTensorImpl(
+      const caffe2::TypeMeta& data_type,
+      c10::Device device_opt)
+      : TensorImpl(TensorTypeId::LazyTensorId, data_type, device_opt),
+        device_(device_opt),
+        dtype_(data_type) {}
+
+  explicit LazyTensorImpl(at::Tensor const& inp)
+      : LazyTensorImpl(inp.dtype(), inp.device()) {
+    inps_.emplace_back(inp);
+  }
+
+  explicit LazyTensorImpl(
+      const char* schema,
+      std::vector<c10::IValue> const& inps,
+      const caffe2::TypeMeta& data_type,
+      c10::Device device_opt)
+      : LazyTensorImpl(data_type, device_opt) {
+    schema_ = schema;
+    for (const auto& inp : inps) {
+      inps_.emplace_back(inp);
+    }
+  }
+
+  c10::intrusive_ptr<TensorImpl> shallow_copy_and_detach(
+      const c10::VariableVersion& version_counter,
+      bool allow_tensor_metadata_change) const override {
+    auto impl =
+        c10::make_intrusive<LazyTensorImpl>(schema_, inps_, dtype(), device());
+    return impl;
+  }
+
+  const char* schema_ = nullptr;
```

Comment:
It looks like in some cases, `schema_ == nullptr`. When this is the case, what does the lazy tensor semantically represent? It kinds of sounds like you have an invariant like "if schema is null, then `inps_.size() == 1`". Is this true?

Reply:
Based on further reading, I think this invariant holds? It would be clearer if there were a more semantic function we use to test for this case, rather than manually always poking to check if `schema_` is null